### PR TITLE
Some PyTableModel & PyListModel touch-ups

### PR DESCRIPTION
**Pull request by kernc**
These are actually pretty joyful to use. We should do this more often. :smiley:

### CODE REVIEW

**Comment by kernc** [comment]
@matevzkren, @PrimozGodec Could one or preferably both of you guys take a look at and review this PR. Consider it mandatory part of your assignments. You both will also have tableviews to work with eventually, I suppose.

File: Orange/widgets/utils/itemmodels.py [hunk 1]
```diff
@@ -76,11 +85,29 @@ def columnCount(self, parent=QModelIndex()):
...
+                absval = abs(value)
+                strlen = len(str(int(absval)))
+                value = '{:.{}{}}'.format(value,
+                                          2 if absval < .001 else
+                                          3 if strlen < 2 else
+                                          1 if strlen < 5 else
+                                          0 if strlen < 6 else
+                                          3,
+                                          'f' if strlen < 6 and absval > .001 else 'e')
+            return str(value)
```
Comment (janezd) [inline Orange/widgets/utils/itemmodels.py]: `value` may already be a string here. Perhaps `return` inside `if`?

On a related note, I'm concerned about the speed of this conversion if the table is huge.

File: Orange/widgets/utils/itemmodels.py [hunk 2]
```diff
@@ -76,11 +85,29 @@ def columnCount(self, parent=QModelIndex()):
...
+            if (isinstance(value, Number) and
+                    not (isnan(value) or isinf(value) or isinstance(value, Integral))):
+                absval = abs(value)
+                strlen = len(str(int(absval)))
+                value = '{:.{}{}}'.format(value,
+                                          2 if absval < .001 else
+                                          3 if strlen < 2 else
+                                          1 if strlen < 5 else
+                                          0 if strlen < 6 else
+                                          3,
```
Comment (janezd) [inline Orange/widgets/utils/itemmodels.py]: These if's are difficult to follow, at least for me. Is this perhaps cleaner?
```
2 if absval < 0.001 else
1 if 2 <= strlen <= 4 else
0 if strlen == 5 else
3
```

File: Orange/widgets/utils/itemmodels.py [hunk 3]
```diff
@@ -145,6 +154,7 @@ def sort(self, column, order=Qt.AscendingOrder):
                          key=lambda i: self._table[i][column],
                          reverse=order != Qt.AscendingOrder)
         self._table[:] = [self._table[i] for i in indices]
+        self._roleData[:] = [self._roleData[i] for i in indices]
```
Comment (janezd) [inline Orange/widgets/utils/itemmodels.py]: I may have missed something, but what if `_roleData` is `None`? Same in other methods that follow.

Comment (kernc) [inline Orange/widgets/utils/itemmodels.py]: Possibly. Where would it be set to `None`?

File: Orange/widgets/utils/itemmodels.py (hunk 1 above)
Comment (kernc) [inline Orange/widgets/utils/itemmodels.py]: No.

File: Orange/widgets/utils/itemmodels.py (hunk 3 above)
Comment (janezd) [inline Orange/widgets/utils/itemmodels.py]: As I said, I may have missed something. I overlooked that the constructor calls `wrap`, and that calling `wrap` is also the only (legal) way of assigning to `_table`.

File: Orange/widgets/utils/itemmodels.py [hunk 4]
```diff
@@ -222,39 +239,34 @@ def __setitem__(self, i, value):
     def wrap(self, table):
         self.beginResetModel()
         self._table = table
+        rows, cols = range(len(table)), range(len(table[0]) if table else 0)
+        self._roleData = [[{} for _ in cols] for _ in rows]
```
Comment (janezd) [inline Orange/widgets/utils/itemmodels.py]: I'm not sure this is a good idea. An empty dictionary takes 288 bytes on my system. A very modest table with 10 columns and 100 rows thus takes 288 KB for empty dicts. 10000 columns and 100 rows -- which is still not outrageously big -- will take 288.000,000 bytes. If `table` is, say, a numpy array of doubles, the data will take 8 MB and the `_roleData`, which is create automatically, just for the case we might need it, will take 288 MB.

Comment (kernc) [inline Orange/widgets/utils/itemmodels.py]: I was following PyListModel example (except where noticeably cleaner). If 100,000,000 rows are added to PyListModel, its `_other_data` grows to overwhelming 29 GB.

Comment (janezd) [inline Orange/widgets/utils/itemmodels.py]: I don't know whether this was sarcasm or not. To be on the safe side, I'll reply. :)

It is unlikely that `PyListModel` will have 1e8 rows, but quite likely that a table will have 1e7 (=1000 x 10000) cells and take 2,9 GB - without the widget actually using other roles.

We use `PyListModel` for settings or attribute lists, which are reasonably small, so no harm done here. I support the idea of adding other roles, but setting individual cell roles will make sense only for small tables (like confusion matrices, perhaps). If larger tables define roles, they will overload the method `data` and compute the role values on the fly. I guess `_roleData` should be `None` unless set explicitly.

Comment (kernc) [inline Orange/widgets/utils/itemmodels.py]: PyTableModel was, likewise, meant for smaller tables like confusion matrices, Test & Score results etc. But given how useful it appears to be, I'll take you up on it and make it even better.

**Pull request merged**